File: Fraction to Recurring Decimal.java
public class Solution {
    public String fractionToDecimal(int _numerator, int _denominator) {
        if (_numerator == 0)
            return "0";
            
        StringBuilder decimal = new StringBuilder();
        boolean plus = (_numerator >= 0 && _denominator >= 0) || (_numerator <= 0 && _denominator <= 0);
        if (!plus)
            decimal.append('-');
        long numerator = Math.abs((long)_numerator);
        long denominator = Math.abs((long)_denominator);
        long quotient = numerator / denominator;
        decimal.append(Long.toString(quotient));
        long reminder = numerator % denominator;
        if (reminder == 0)
            return decimal.toString();
        decimal.append('.');
        Map<Long, Integer> cache = new HashMap<Long, Integer>();
        List<Long> digits = new ArrayList<Long>();
        int index = -1;
        while (reminder > 0) {
            numerator = reminder;
            if (cache.containsKey(numerator)) {
                index = cache.get(numerator);
                break;
            }
            cache.put(numerator, digits.size());
            numerator *= 10;
            quotient = numerator / denominator;
            digits.add(quotient);
            reminder = numerator % denominator;
        } 
        for (int i = 0; i < digits.size(); i++) {
            if (i == index)
                decimal.append('(');
            decimal.append((char)('0' + digits.get(i)));
        }
        if (index != -1)
            decimal.append(')');
        return decimal.toString();
    }
}

File: Fraction to Recurring Decimal.py
class Solution:
	# @return a string
	def fractionToDecimal(self, numerator, denominator):
		if (numerator == 0):
			return "0"
		ret = ""
		if ((numerator > 0 and denominator < 0) or (numerator < 0 and denominator > 0)):
			ret = "-"
		numerator = abs(numerator)
		denominator = abs(denominator)
		quotient = numerator / denominator
		ret = ret + str(quotient)
		reminder = numerator % denominator
		if (reminder == 0):
			return ret
		ret = ret + '.'
		cache = dict()
		digits = []
		index = -1
		while (reminder > 0):
			numerator = reminder
			if (cache.has_key(numerator)):
				index = cache[numerator]
				break
			cache[numerator] = len(digits)
			numerator = numerator * 10
			quotient = numerator / denominator
			reminder = numerator % denominator
			digits.append(quotient)
		for i in xrange(len(digits)):
			if (i == index):
				ret = ret + '('
			ret = ret + str(digits[i])
		if (index != -1):
			ret = ret + ')'
		return ret
